### ISSUE TITLE
Endpoint, Token, Cert Path, and Cert Key Storage for CLI Tool

### ISSUE THREAD

**Issue by tolgaozen** [Permify/permify-cli#2]
### Description
We need to implement a feature for storing the endpoint, token, cert path, and cert key in the CLI tool. These credentials should be retrievable from the storage for later use during the client creation process.

### Location
`permify-cli/internal/client/client.go`

### Proposed Change
Modify the `New` function in `client.go` to retrieve stored credentials. The updated function could look something like this:

```go
func New(endpoint string) (*permify.Client, error) {
    // Retrieve credentials from storage
    // ... (code to retrieve credentials)

    client, err := permify.NewClient(
        permify.Config{
            Endpoint: endpoint,
            // Other necessary configurations
        },
        grpc.WithTransportCredentials(insecure.NewCredentials()),
    )
    return client, err
}
```

### Additional Notes
Ensure that the credential retrieval process is secure and efficient.
Include error handling for cases where credentials are missing or incorrect.

**Comment by Pavan-Nambi** [comment]
hey @tolgaOzen  can i work on this?

**Comment by Pavan-Nambi** [comment]
hey @tolgaOzen  , I am bit new to codebase but by storage u mean to store them in some env/yml files? or something else?

**Comment by tolgaOzen** [comment]
Hello @Pavan-Nambi, by storage I mean a way to prevent users from having to repeatedly enter their credentials when connecting to the Permify API. For this, you can use the path ~/.permify/credentials on macOS and Linux, and C:\Users\USERNAME\.permify\credentials on Windows. Here, you can store the endpoint, token, certificate path, and certificate key. When user wants to communicate with Permify APIs, Permify CLI can read these details from there. I hope this is clear.

**Comment by Pavan-Nambi** [comment]
got it, thanks

**Comment by Pavan-Nambi** [comment]
@tolgaOzen  i've made a pr #8 based on what i understood  lmk if i did anything wrong, or if u want me to change anything

**Comment by asalvi0** [comment]
Putting myself on the queue for collaboration or in case this needs to be reassigned ✋

I think it requires storing secrets securely. Also, I have a couple of follow-up questions:

* will there be a follow-up task to actually use the retrieved credentials while building the client, or is expected to be in this one?
* **what's the expected mechanism to allow the user changing these values?**, and will there be a follow-up task for this?

**Comment by Rutik7066** [comment]
Hey @tolgaOzen can i work on this one?

**Comment by vivasvan1** [comment]
Putting myself on the queue for collab or in case this needs to be reassigned ✋
@tolgaOzen

**Comment by snigdha510** [comment]
Putting myself on the queue for collab or in case this needs to be reassigned ✋
@tolgaOzen

**Comment by kumarvikramshahi** [comment]
Since the issue is not assigned to anyone, I would like to give it a shot or would be happy to collab if possible. 
@tolgaOzen  Please check if the issue can be reassigned. 
/attempt #2

<div id="algora-attempt" />

<details>
  <summary>Options</summary>
  <ul>
    <li>
      <a href="https://console.algora.io/api/bounties/clpwe63nl0002jk0ft1tzsviu/cancel-attempt">
        Cancel my attempt
      </a>
    </li>
  </ul>
</details>

**Comment by connorbrez** [comment]
Hello @tolgaOzen, it doesn't look like anyone is assigned this task. I am interested in working on this, could you assign me this task?

**Comment by Sadaf-A** [comment]
/attempt #2 

<div id="algora-attempt" />

<details>
  <summary>Options</summary>
  <ul>
    <li>
      <a href="https://console.algora.io/api/bounties/clpwe63nl0002jk0ft1tzsviu/cancel-attempt">
        Cancel my attempt
      </a>
    </li>
  </ul>
</details>

**Comment by Sadaf-A** [comment]
Greetings @tolgaOzen, could you please assign this to me?

**Comment by radasam** [comment]
/attempt #2

<div id="algora-attempt" />

<details>
  <summary>Options</summary>
  <ul>
    <li>
      <a href="https://console.algora.io/api/bounties/clpwe63nl0002jk0ft1tzsviu/cancel-attempt">
        Cancel my attempt
      </a>
    </li>
  </ul>
</details>

**Comment by radasam** [comment]
@tolgaOzen I can see that since this was created a config file has been implemented to store the endpoint, Do we just need to follow the same process for the token + cert, or is this not needed anymore?

**Comment by akhilender-bongirwar** [comment]
@tolgaOzen, is this issue still open to work?

**Comment by nicefacer** [comment]
@tolgaozen review PR and get back to me.

**Comment by CuboYe** [comment]
/claim 2

Claiming territory for a scoped fix attempt now. If maintainer preference differs, I can adapt quickly.

**Comment by mitchellecm7** [comment]
Questions before I start on #2

I've built CLI tools with credential storage before (keyring, encrypted local config, etc.). A few things I'd want to clarify before writing code:

1. Where should credentials live?

OS keyring (Keychain/Secret Service/Credential Manager)?

Encrypted file in ~/.permify/config?

Plain text (please no)?

2. What's the auth flow?

Does endpoint + token cover all cases, or is cert path/key always required too?

Are there scenarios where you'd have endpoint + token but no certs (or vice versa)?

3. Storage format

Multiple profiles? (e.g., --profile prod vs --profile dev)

If yes, we need profile name as part of the lookup key

4. CLI commands needed?

permify config set --endpoint https://... --token xxx

permify config list

permify config use --profile prod

Or are you thinking purely programmatic storage with no user-facing commands yet?

My recommendation if you want a minimal first version:

Encrypted JSON file at $XDG_CONFIG_HOME/permify/credentials.json (AES-GCM with a key derived from a user-provided passphrase, or just stored in OS keyring)

Single profile initially, expand to multiple later

New() reads from that file; if missing, returns a clear error like "no stored credentials found, run permify config init"

**Comment by mitchellecm7** [comment]
hey @tolgaozen can i work on this?

**Comment by Leviann** [comment]
/attempt #2

Plan (already shipped in PR #20):
- Extend profile YAML with token, cert_path, cert_key_path; wire configure prompts and gRPC dial (mTLS + bearer).
- NewFromCLIConfig for cmd clients; Unix chmod 0600 when any profile stores a token.
- Tests: grpc normalization and dial-option validation; config token detection and Unix perm (non-Windows build tag).

**Comment by paulco85** [comment]
Hi, is this bounty still open for a fresh contributor? I see multiple existing PRs and Algora's note about assignment-first, so I'd like to confirm before starting. If you still want this implemented or reworked, I can prepare a small scoped PR following the requested credentials file format.

**Comment by saiprasadchary** [comment]
/attempt #2

Implementation plan:
- keep the existing profile config for tenant/profile selection, but move connection credentials into ~/.permify/credentials as requested in the maintainer note
- store endpoint, token, cert path, and cert key path per profile in that credentials file so existing --profile workflows still work
- update permctl configure to prompt for those values, validate the connection with them, and persist them separately from the normal config YAML
- update the gRPC client to load bearer token and optional mTLS client cert/key from the stored credentials when commands create a client
- add focused tests for credentials path handling, config/credential loading, and token/TLS client setup

Assumption: using the existing profile name as the key inside ~/.permify/credentials is acceptable so the change stays backward-compatible with current multi-profile CLI usage.

**Comment by mike123456789009** [comment]
Hi, I did a local implementation spike for this before opening another PR because I saw several open credential/config PRs and a fresh `/attempt` today.

My local branch adds:
- profile-aware credential storage at `~/.permify/credentials`
- endpoint, token, cert path, and cert key persistence
- `0600` writes for the credentials file
- `configure` prompts to collect/persist the credential fields
- gRPC client loading stored credentials, bearer metadata, and TLS cert/key when configured
- tests for credential read/write and client credential loading

Verification passes locally: `go test ./...`, `go vet ./...`, `go build ./cmd/permctl`, and `git diff --check`.

Before I open a PR and claim the bounty, can a maintainer confirm whether a fresh implementation in this shape is still wanted, or whether one of the existing PRs is already the preferred direction?

**Comment by saiprasadchary** [comment]
Implemented in #25.

This version follows the maintainer-requested `~/.permify/credentials` storage path instead of persisting token and cert settings in the normal profile YAML, while keeping the existing profile/tenant workflow intact.

Validation:
- `/tmp/go/bin/go test ./...`

**Comment by devz0x** [comment]
Hey folks I’d love to take this bounty if it’s still up for grabs.

I can wire up credential storage/retrieval for endpoint, token, cert path, and cert key, plus tests and proper error handling (because “it works on my machine” is not a security model).

**Comment by asim48-ctrl** [comment]
I can implement credential persistence for endpoint/token/cert paths.

Proposed approach: add a small config store (OS keychain if desired, otherwise file-based in `~/.config`), add `permify config set/get` commands, and update `internal/client/client.go` to load defaults when args aren’t provided.

Any preference on storage mechanism (keyring vs plain config file) and on overriding precedence (CLI flags > env vars > config > prompt)?

**Comment by TheDerbiedOne** [comment]
Implemented persistent endpoint/token/cert storage + client wiring. PR: https://github.com/Permify/permify-cli/pull/29

**Comment by sercondd-glitch** [comment]
redy to get this assigned @tolgaozen

**Comment by minnhnh** [comment]
Opened PR #34 to implement persistent storage/retrieval of token + mTLS cert paths (and env overrides), as requested.\n\nPR: https://github.com/Permify/permify-cli/pull/34

**Comment by watcharaponthod-code** [comment]
/attempt #2

Implemented in PR #35.

Approach follows the maintainer's notes: stores 	oken, cert_path, and cert_key_path in the existing profile YAML. permctl configure prompts for the new fields (all optional). The gRPC client wires bearer auth and mTLS at dial time. PERMIFY_TOKEN, PERMIFY_CERT_PATH, and PERMIFY_CERT_KEY_PATH env vars override stored values for CI/container use. Existing config files without the new keys are unaffected.

**Comment by kitwongpixel** [comment]
Implemented in PR #36: https://github.com/Permify/permify-cli/pull/36

**Comment by tanayv77** [comment]
Opened PR #38 with the config storage, token, and TLS client changes for this issue. It includes regression tests and passes `go test ./...` locally.

**Comment by tanayv77** [comment]
/attempt #2

Updated PR #38 to follow the maintainer-requested dedicated credentials path (~/.permify/credentials) rather than persisting endpoint/token/cert material in the main profile YAML. Validation still passes with go test ./....

**Comment by tanayv77** [comment]
/claim #2

Tracking in PR #38: https://github.com/Permify/permify-cli/pull/38

**Comment by admac-hub** [comment]
Hi! I'm interested in working on this issue.

I have experience with authentication systems, infrastructure automation, and credential management. I've reviewed the issue and would be happy to implement the proposed solution.

Could I be assigned to this one?

**Comment by 1663580752-wq** [comment]
Hi, is this Algora bounty still available for a new implementation PR?

I can work on storing endpoint, token, cert path, and cert key for the CLI and wiring those values into client creation. Before starting, I want to confirm this issue is still open for external contributors.

**Comment by defgtyg** [comment]
/attempt #2

Implemented in PR #43: https://github.com/Permify/permify-cli/pull/43

**Comment by franklincg** [comment]
Hi! I’m interested in working on this bounty. I saw that the project prefers assigning a single contributor before implementation. Is the $200 bounty still active, and is the issue currently available to be assigned to a new contributor?
